[Navigation API] Implement BFCache support for navigation.entries().

basuke opened the pull request:
#### 1623b8c4b1098bff857dff660fa99e750ec026f3
<pre>
[Navigation API] Implement BFCache support for navigation.entries().
<a href="https://bugs.webkit.org/show_bug.cgi?id=297291">https://bugs.webkit.org/show_bug.cgi?id=297291</a>
<a href="https://rdar.apple.com/158164248">rdar://158164248</a>

Reviewed by Alex Christensen.

Implement proper Back-Forward Cache (BFCache) support for the Navigation API&apos;s
navigation.entries() method. Previously, pages with Navigation API were unable
to enter BFCache, and iframe Navigation APIs were not updated during restoration.

This change enables Navigation API pages to be cached and ensures that
navigation.entries() is properly updated when restored from BFCache, including
correct disposal of unreachable forward history entries.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-bfcache.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-after-bfcache.html:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::reachableItemsForFrame):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):

Canonical link: <a href="https://commits.webkit.org/298691@main">https://commits.webkit.org/298691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eec4d5968165cd765b4f32ed0aee214260f9c359

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116369 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36030 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26580 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122426 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/66930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118258 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36724 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44618 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88385 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/66930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119318 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29287 "1 flakes") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104399 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68818 "Passed tests") | | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28370 "Exiting early after 10 failures. 43 tests run. 1 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22502 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66107 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98669 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22663 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125575 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43263 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32488 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97092 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43628 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100605 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/96887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24651 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/42176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20074 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39217 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43151 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48742 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42617 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/45957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44322 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

basuke commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -186,13 +186,34 @@ void CachedPage::restore(Page& page)
-    firePageShowEvent(page);
+        // Update Navigation API for all child frames (iframes)
+        Vector<Ref<LocalFrame>> childFrames;
+        for (auto* child = localMainFrame->tree().traverseNextInPostOrder(CanWrap::Yes); child; child = child->tree().traverseNextInPostOrder(CanWrap::No)) {
```
I have no plan for RemoteFrame! Let me think if first.

achristensen07 commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -186,13 +185,34 @@ void CachedPage::restore(Page& page)
+        }
+
+        for (Ref child : childFrames) {
+            if (!child->tree().isDescendantOf(localMainFrame.get()))
```
I don't see how this could be possible.  Can it be removed?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/19e2a058afe36b1d0dd40cc04bd3efb26ace0703. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49289

basuke committed:
[Navigation API] Implement BFCache support for navigation.entries().
https://bugs.webkit.org/show_bug.cgi?id=297291
rdar://158164248

Reviewed by Alex Christensen.

Implement proper Back-Forward Cache (BFCache) support for the Navigation API's
navigation.entries() method. Previously, pages with Navigation API were unable
to enter BFCache, and iframe Navigation APIs were not updated during restoration.

This change enables Navigation API pages to be cached and ensures that
navigation.entries() is properly updated when restored from BFCache, including
correct disposal of unreachable forward history entries.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-history-back-bfcache.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-bfcache.html:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-after-bfcache-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-after-bfcache.html:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::reachableItemsForFrame):
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):

Canonical link: https://commits.webkit.org/298691@main

webkit-commit-queue commented:
Committed 298691@main (1623b8c4b109): <https://commits.webkit.org/298691@main>

Reviewed commits have been landed. Closing PR #49289 and removing active labels.
